File: commands/utility/wolfram.go
package utility

import (
  "github.com/sapphire-cord/sapphire"
  "github.com/pollen5/taiga/utils"
  "net/http"
  "net/url"
  "io/ioutil"
  "encoding/json"
)

// Query Wolfram Alpha with any mathematical question.
// Usage: <query:string...>
// Aliases: what
func Wolfram(ctx *sapphire.CommandContext) {
  input := ctx.JoinedArgs()

  // If the "what" alias was used treat it as an argument aswell.
  if ctx.InvokedName == "what" {
    input = "What " + input
  }

  query := url.Values{}
  query.Add("input", input)
  query.Add("primary", "true")
  query.Add("appid", utils.GetConfig("wolfram"))
  query.Add("output", "json")

  res, err := http.Get("http://api.wolframalpha.com/v2/query?" + query.Encode())

  if err != nil {
    ctx.Error(err)
    return
  }

  defer res.Body.Close()

  if res.StatusCode != 200 {
    ctx.Error("Wolfram returned non-200 status: %d (%s)", res.StatusCode, res.Status)
    return
  }

  var data struct {
    Result struct {
      Pods []struct {
        SubPods []struct {
          Text string `json:"plaintext"`
        } `json:"subpods"`
      } `json:"pods"`
    } `json:"queryresult"`
  }

  buf, err := ioutil.ReadAll(res.Body)

  if err != nil {
    ctx.Error(err)
    return
  }

  err = json.Unmarshal(buf, &data)

  if err != nil {
    ctx.Error(err)
    return
  }

  if len(data.Result.Pods) < 2 {
    ctx.Reply("Couldn't find an answer to that question!")
    return
  }

  answer := data.Result.Pods[1].SubPods[0].Text

  if len(answer) > 1500 {
    answer = answer[:1500]
  }

  ctx.BuildEmbed(sapphire.NewEmbed().
    SetTitle(data.Result.Pods[0].SubPods[0].Text).
    SetDescription(answer).
    SetColor(0xDFAC7C))
}

File: commands/utility/wolfram.js
const Command = require("../../structures/Command.js");
const { MessageEmbed } = require("discord.js");
const fetch = require("node-fetch");

class Wolfram extends Command {
  constructor(...args) {
    super(...args, {
      description: "Query Wolfram Alpha with any mathematical question.",
      usage: "wolfram <query>",
      aliases: ["what", "when", "where", "who", "why", "how"],
      cooldown: 5
    });
  }
  
  async run(ctx, args) {
    if(!args.length) return ctx.reply("What are you trying to ask?");
    
    // Allow users to trigger this in a fancy way using @Miyako What time is it?
    // If they invoke it with the "what"/"where"/"when"/"who"/"why" alias, we must also treat it as an argument.
    // A trick to make it look like some advanced A.I bot i guess.
    const query = ctx.invokedName === "what" || ctx.invokedName === "when" || ctx.invokedName === "how" ||
      ctx.invokedName === "where" || ctx.invokedName === "who" || ctx.invokedName === "why" ?
      `${this.client.utils.toProperCase(ctx.invokedName)} ${args.join(" ")}` : args.join(" ");

    const url = new URL("http://api.wolframalpha.com/v2/query");
    url.search = new URLSearchParams([
      ["input", query],
      ["primary", true],
      ["appid", this.client.config.wolfram],
      ["output", "json"]
    ]);
    
    const pods = await fetch(url)
      .then((res) => res.json())
      .then((body) => body.queryresult.pods);
    
    if(!pods || pods.error) return ctx.reply("Couldn't find an answer to that question!");

    return ctx.reply(new MessageEmbed()
      .setTitle(pods[0].subpods[0].plaintext)
      .setDescription(pods[1].subpods[0].plaintext.substring(0, 1950))
      .setColor(0x9590EE));
  }
}

module.exports = Wolfram;
